x-pack/plugins/rule_registry/server/rule_registry/rule_type_helpers/create_lifecycle_rule_type_factory.ts

```diff
@@ -0,0 +1,235 @@
+import { ActionVariable, AlertInstanceState } from '../../../../alerting/common';
+import { RuleParams, RuleType } from '../../types';
+import { DefaultFieldMap } from '../defaults/field_map';
+import { OutputOfFieldMap } from '../field_map/runtime_type_from_fieldmap';
+import { PrepopulatedRuleEventFields } from '../create_scoped_rule_registry_client/types';
+import { RuleRegistry } from '..';
+
+type UserDefinedAlertFields<TFieldMap extends DefaultFieldMap> = Omit<
+  OutputOfFieldMap<TFieldMap>,
+  PrepopulatedRuleEventFields | 'kibana.rac.alert.id' | 'kibana.rac.alert.uuid' | '@timestamp'
+>;
+
+type LifecycleAlertService<
+  TFieldMap extends DefaultFieldMap,
+  TActionVariable extends ActionVariable
+> = (alert: {
+  id: string;
+  fields: UserDefinedAlertFields<TFieldMap>;
+}) => AlertInstance<AlertInstanceState, { [key in TActionVariable['name']]: any }, string>;
+
+type CreateLifecycleRuleType<TFieldMap extends DefaultFieldMap> = <
+  TRuleParams extends RuleParams,
+  TActionVariable extends ActionVariable
+>(
+  type: RuleType<
+    TFieldMap,
+    TRuleParams,
+    TActionVariable,
+    { alertWithLifecycle: LifecycleAlertService<TFieldMap, TActionVariable> }
+  >
+) => RuleType<TFieldMap, TRuleParams, TActionVariable>;
+
+const trackedAlertStateRt = t.type({
+  alertId: t.string,
+  alertUuid: t.string,
+  started: t.string,
+});
+
+const wrappedStateRt = t.type({
+  wrapped: t.record(t.string, t.unknown),
+  trackedAlerts: t.record(t.string, trackedAlertStateRt),
+});
+
+export function createLifecycleRuleTypeFactory<
+  TRuleRegistry extends RuleRegistry<DefaultFieldMap>
+>(): TRuleRegistry extends RuleRegistry<infer TFieldMap>
+  ? CreateLifecycleRuleType<TFieldMap>
+  : never;
+
+export function createLifecycleRuleTypeFactory(): CreateLifecycleRuleType<DefaultFieldMap> {
+  return (type) => {
+    return {
+      ...type,
+      executor: async (options) => {
+        const {
+          services: { scopedRuleRegistryClient, alertInstanceFactory, logger },
+          state: previousState,
+          rule,
+        } = options;
+
+        const decodedState = wrappedStateRt.decode(previousState);
+
+        const state = isLeft(decodedState)
+          ? {
+              wrapped: previousState,
+              trackedAlerts: {},
+            }
+          : decodedState.right;
+
+        const currentAlerts: Record<
+          string,
+          UserDefinedAlertFields<DefaultFieldMap> & { 'kibana.rac.alert.id': string }
+        > = {};
+
+        const timestamp = options.startedAt.toISOString();
+
+        const nextWrappedState = await type.executor({
+          ...options,
+          state: state.wrapped,
+          services: {
+            ...options.services,
+            alertWithLifecycle: ({ id, fields }) => {
+              currentAlerts[id] = {
+                ...fields,
+                'kibana.rac.alert.id': id,
+              };
+              return alertInstanceFactory(id);
```

Comment:
I'm curious to see how the alerting notion of an "alert instance id" works out here.  I know we've been talking about an "alert id", which would identify a "span" (not sure of the current terminology) of a number of consecutive alerts being triggered.  We don't have such a thing in alerting today, but would be straight-forward to create UUIDs for these.

I think it may work out well to keep this as an rule-specific thing, so that for rules that care about these "spans", they'd need to make sure they provide unique `id` values when appropriate, but for rules that don't, they can reuse `id` values as is often done in rules today (for example, an `id` _could_ identify a term used in grouping for an agg, which would not be unique over time).

Assuming we eventually add an automagic UUID to identify a span of consecutive alerts, the rule registry layer could make use of that directly.